stablehlo/dialect/TypeInference.cpp

```diff
@@ -3366,6 +3368,80 @@ LogicalResult verifyCollectivePermuteOp(
   return success();
 }
 
+LogicalResult verifyCompositeOp(std::optional<Location> loc, Operation* op,
+                                StringRef name, StringRef decomposition) {
+  // composite_c1
+  llvm::SmallVector<llvm::StringRef> parts;
+  name.split(parts, '.');
+  if (
+      // Name must have at least 2 non-empty parts separated by periods.
+      parts.size() < 2 ||
+      llvm::any_of(parts, [](auto part) { return part.empty(); }) ||
+      // First part's first character must be alpha, and subsequent characters
+      // must be alpha, digit or undescore.
+      !isalpha(parts[0][0]) ||
+      !llvm::all_of(
+          llvm::drop_begin(parts[0]),
+          [](char c) { return isalpha(c) || isdigit(c) || c == '_'; }) ||
+      // Subsequent parts must be all alpha, digit, underscore or dollar sign.
+      !llvm::all_of(llvm::drop_begin(parts), [](llvm::StringRef part) {
+        return llvm::all_of(part, [](char c) {
+          return isalpha(c) || isdigit(c) || c == '_' || c == '$';
+        });
+      })) {
+    return emitOptionalError(loc,
```

Comment:
personal preference: break these out into multiple conditions and you can give an error message for the more meaningful failure.

Reply:
I tried breaking it out and I think the above is still preferable. I agree that by breaking it out we can give more specific error messages (e.g. "expected at least 2 non-empty parts separated by periods, but got: ..."), but IMO giving the full expected pattern in the error message makes it easiest to know what needs to be done to get to a valid name. In general I'm not a huge fan of massive conditions like this one but I think this one reads fine, it reads kind of like a regular expression thanks to the LLVM helpers.